clang/include/clang/CIR/Dialect/IR/CIROps.td

```diff
@@ -1630,6 +1635,52 @@ def GetMemberOp : CIR_Op<"get_member"> {
   let hasVerifier = 1;
 }
 
+//===----------------------------------------------------------------------===//
+// VecElemOp
+//===----------------------------------------------------------------------===//
+
+def VecElemOp : CIR_Op<"vec_elem", [Pure,
+  TypesMatchWith<"type of 'result' matches element type of 'vec'",
+                 "vec", "result",
+                 "$_self.cast<VectorType>().getEltType()">]> {
+
+  let summary = "Extract one element from a vector object";
+  let description = [{
+    The `cir.vec_elem` operation extracts one element from a vector object.
+  }];
+
+  let arguments = (ins CIR_VectorType:$vec, CIR_IntType:$index);
+  let results = (outs AnyType:$result);
+
+  let assemblyFormat = [{
+    $vec `[` $index `:` type($index) `]` type($vec) `->` type($result) attr-dict
+  }];
+
+  let hasVerifier = 0;
+}
+
+//===----------------------------------------------------------------------===//
+// VecValue
+//===----------------------------------------------------------------------===//
+
+def VecValueOp : CIR_Op<"vec", [Pure]> {
+
+  let summary = "Create non-const vector value";
+  let description = [{
+    The `cir.vec` operation creates a vector value with the given element
+    values.
+  }];
+
+  let arguments = (ins Variadic<AnyType>:$elements);
+  let results = (outs CIR_VectorType:$result);
+
+  let assemblyFormat = [{
+    `(` ($elements^ `:` type($elements))? `)` `:` type($result) attr-dict
```

Comment:
What does it mean to have no elements?

Reply:
A vector with no elements isn't possible.  The front end will reject it before ClangIR is ever generated.  Is there a way to express that constraint in the `.td` files?  (TableGen is new to me.  I still have a lot to learn about how best to use it.)

Reply:
It's possible you could write a constraint similar to `UInt` and friends in `CIRTypes.td` (see how they are used in `cir.lib.memchr`). The generated .cpp's could provide some insights on how to deal with `Variadic` when writing the constrains. If it's too hard to write, you could just make the op verifier blow on zero sized elements.

Reply:
I spent some time trying to add more validation checks to TableGen.  But I am struggling with that at the moment.  I plan to come back and clean up some of the TableGen definitions a little later once I have had more time to figure out how it works.  I don't want to hold up this PR any longer for this, since what is here works, if not perfect yet.